tables scanning for orderedness when they shouldn't be

example:

```
In [10]: hl.utils.range_table(1).select(x=[1])
2018-05-02 23:41:10 Hail: INFO: Coerced sorted dataset
Out[10]: <hail.table.Table at 0x108dffcc0>
```
Note that if you annotate `x=1`, it doesn't scan (this is related to broadcasts or something)

@patrick-schultz Is the bug from this warning?

```
[ant:scalac] /Users/jigold/hail/src/main/scala/is/hail/table/Table.scala:584: warning: comparing values of types Array[String] and Int using `==' will always yield false
[ant:scalac]               ordered.typ.partitionKey == partitionKeys.length) {
```

--- a/src/main/scala/is/hail/expr/Relational.scala
+++ b/src/main/scala/is/hail/expr/Relational.scala
@@ -1451,6 +1451,7 @@ case class TableImport(paths: Array[String], typ: TableType, readerOpts: TableRe
 }
 
 case class TableRange(n: Int, nPartitions: Int) extends TableIR {
+  private val nPartitionsAdj = math.min(n, nPartitions)
   val children: IndexedSeq[BaseIR] = Array.empty[BaseIR]
 
   def copy(newChildren: IndexedSeq[BaseIR]): TableRange = {
@@ -1458,7 +1459,7 @@ case class TableRange(n: Int, nPartitions: Int) extends TableIR {
     this
   }
 
-  private val partCounts = partition(n, nPartitions)
+  private val partCounts = partition(n, nPartitionsAdj)
 
   override val partitionCounts = Some(partCounts.map(_.toLong))
 
@@ -1477,12 +1478,12 @@ case class TableRange(n: Int, nPartitions: Int) extends TableIR {
       new OrderedRVD(
         new OrderedRVDType(Array("idx"), Array("idx"), typ.rowType),
         new OrderedRVDPartitioner(Array("idx"), typ.rowType,
-          Array.tabulate(nPartitions) { i =>
+          Array.tabulate(nPartitionsAdj) { i =>
             val start = partStarts(i)
-            val end = partStarts(i+1)
+            val end = partStarts(i + 1)
             Interval(Row(start), Row(end), includesStart = true, includesEnd = false)
           }),
-        hc.sc.parallelize(Range(0, nPartitions), nPartitions)
+        hc.sc.parallelize(Range(0, nPartitionsAdj), nPartitionsAdj)
           .mapPartitionsWithIndex { case (i, _) =>
             val region = Region()
             val rvb = new RegionValueBuilder(region)

--- a/src/main/scala/is/hail/expr/Relational.scala
+++ b/src/main/scala/is/hail/expr/Relational.scala
@@ -1474,7 +1474,14 @@ case class TableRange(n: Int, nPartitions: Int) extends TableIR {
 
     TableValue(typ,
       BroadcastRow(Row(), typ.globalType, hc.sc),
-      new UnpartitionedRVD(typ.rowType,
+      new OrderedRVD(
+        new OrderedRVDType(Array("idx"), Array("idx"), typ.rowType),
+        new OrderedRVDPartitioner(Array("idx"), typ.rowType,
+          Array.tabulate(nPartitions) { i =>
+            val start = partStarts(i)
+            val end = partStarts(i+1)
+            Interval(Row(start), Row(end), includesStart = true, includesEnd = false)
+          }),
         hc.sc.parallelize(Range(0, nPartitions), nPartitions)
           .mapPartitionsWithIndex { case (i, _) =>
             val region = Region()

--- a/src/main/scala/is/hail/table/Table.scala
+++ b/src/main/scala/is/hail/table/Table.scala
@@ -47,10 +47,8 @@ case class TableSpec(
 }
 
 object Table {
-  def range(hc: HailContext, n: Int, nPartitions: Option[Int] = None): Table = {
-    require(nPartitions.forall(_ <= n))
-    new Table(hc, TableRange(n, nPartitions.getOrElse(math.min(n, hc.sc.defaultParallelism))))
-  }
+  def range(hc: HailContext, n: Int, nPartitions: Option[Int] = None): Table =
+    new Table(hc, TableRange(n, nPartitions.getOrElse(hc.sc.defaultParallelism)))
 
   def fromDF(hc: HailContext, df: DataFrame, key: java.util.ArrayList[String]): Table = {
     fromDF(hc, df, if (key == null) None else Some(key.asScala.toArray.toFastIndexedSeq))

--- a/src/main/scala/is/hail/table/Table.scala
+++ b/src/main/scala/is/hail/table/Table.scala
@@ -47,8 +47,10 @@ case class TableSpec(
 }
 
 object Table {
-  def range(hc: HailContext, n: Int, nPartitions: Option[Int] = None): Table =
-    new Table(hc, TableRange(n, nPartitions.getOrElse(hc.sc.defaultParallelism)))
+  def range(hc: HailContext, n: Int, nPartitions: Option[Int] = None): Table = {
+    require(nPartitions.forall(_ <= n))
+    new Table(hc, TableRange(n, nPartitions.getOrElse(math.min(n, hc.sc.defaultParallelism))))
+  }
 
   def fromDF(hc: HailContext, df: DataFrame, key: java.util.ArrayList[String]): Table = {
     fromDF(hc, df, if (key == null) None else Some(key.asScala.toArray.toFastIndexedSeq))
@@ -581,7 +583,7 @@ class Table(val hc: HailContext, val tir: TableIR) {
       val orvd = rvd match {
         case ordered: OrderedRVD =>
           if (keys.zip(ordered.typ.key).forall{ case (l, r) => l == r } &&
-              ordered.typ.partitionKey == partitionKeys.length) {
+              ordered.typ.partitionKey.length == partitionKeys.length) {
             if (keys.length <= ordered.typ.key.length)
               ordered.copy(typ = ordered.typ.copy(key = keys))
             else if (ordered.typ.key.length <= keys.length) {

